fem/fespace.cpp

```diff
@@ -3936,6 +4253,57 @@ void FiniteElementSpace::Update(bool want_transform)
    }
 }
 
+void FiniteElementSpace::PRefineAndUpdate(const Array<pRefinement> & refs,
+                                          bool want_transfer)
+{
+   MFEM_VERIFY(PRefinementSupported(),
+               "p-refinement is not supported in this space");
+
+   if (want_transfer)
+   {
+      fesPrev.reset(new FiniteElementSpace(mesh, fec));
+      for (int i = 0; i<mesh->GetNE(); i++)
+      {
+         fesPrev->SetElementOrder(i, GetElementOrder(i));
+      }
+      fesPrev->Update(false);
+   }
+
+   for (auto ref : refs)
+   {
+      SetElementOrder(ref.index, GetElementOrder(ref.index) + ref.delta);
+   }
+
+   Update(false);
+
+   if (want_transfer)
+   {
+      PTh.reset(new PRefinementTransferOperator(*fesPrev, *this));
+   }
+
+   lastUpdatePRef = true;
+}
+
+bool FiniteElementSpace::PRefinementSupported()
+{
+   // Check whether the space type is L2 or H1
+   if (!dynamic_cast<const L2_FECollection*>(fec) &&
+       !dynamic_cast<const H1_FECollection*>(fec))
+   {
+      return false;
+   }
+
+   // Check whether the mesh is purely quadrilateral or hexahedral.
+   const int dim = mesh->Dimension();
+   Array<Geometry::Type> geoms;
+   mesh->GetGeometries(dim, geoms);
+   if (geoms.Size() != 1) { return false; }
+   if (dim == 2 && geoms[0] != Geometry::Type::SQUARE) { return false; }
+   else if (geoms[0] != Geometry::Type::CUBE) { return false; }
```

Comment:
This will fail if `dim == 2 && geoms[0] == Geometry::Type::SQUARE`
```suggestion
   else if (dim == 3 && geoms[0] != Geometry::Type::CUBE) { return false; }
```

Reply:
Good catch. This is a recent change made during the review, which has not been tested fully.

Reply:
Fixed in 7fe4c3c19943a1b5064aa33e99b357f41dbb75c7